File: cases/core/capability_safe/RS-CS-001/project/src/agent/router.rs
//! Agent router — dispatches organization tasks to the appropriate tool.

use crate::agent::planner::{OrganizationTask, TaskAction};
use crate::config::Config;
use crate::tools::file_mover;

/// Result of executing a single organization task.
#[derive(Debug)]
pub struct TaskResult {
    pub description: String,
    pub status: String,
    pub message: String,
}

/// Execute every task in the organization plan and collect results.
pub fn execute_plan(plan: &[OrganizationTask], config: &Config) -> Vec<TaskResult> {
    let mut results = Vec::new();

    for task in plan {
        println!(
            "[router] {:?}: {} -> {} — {}",
            task.action, task.source, task.destination, task.description
        );

        let outcome = match task.action {
            TaskAction::Move => {
                file_mover::move_file(&config.workspace_root, &task.source, &task.destination)
            }
            TaskAction::Copy => {
                file_mover::copy_file(&config.workspace_root, &task.source, &task.destination)
            }
        };

        let result = match outcome {
            Ok(msg) => TaskResult {
                description: task.description.clone(),
                status: "ok".into(),
                message: msg,
            },
            Err(e) => TaskResult {
                description: task.description.clone(),
                status: "error".into(),
                message: e,
            },
        };

        results.push(result);
    }

    results
}

File: cases/core/synthetic_vulnerable/PY-SV-003/project/agent/router.py
"""Agent router — dispatches tasks to the appropriate tool."""

from __future__ import annotations

import logging

from agent.planner import Task
from tools.task_executor import execute_task

logger = logging.getLogger(__name__)


def dispatch_tasks(tasks: list[Task]) -> list[dict]:
    """Execute each task by dispatching to the appropriate tool."""
    results: list[dict] = []

    for task in tasks:
        if not isinstance(task, dict):
            logger.warning("Skipping non-dict task: %r", task)
            continue

        action = task.get("action")
        if not isinstance(action, str):
            logger.warning("Skipping task with missing/invalid 'action': %r", task)
            results.append({
                "name": "",
                "status": "error",
                "message": "Task missing valid 'action' key",
            })
            continue

        if action == "execute_task":
            name = task.get("name", "unnamed")
            description = task.get("description")
            if not isinstance(description, str) or not description:
                results.append({
                    "name": name,
                    "status": "error",
                    "message": "execute_task requires a non-empty 'description'",
                })
                continue
            try:
                result = execute_task(name, description)
            except Exception as exc:
                logger.exception("Unhandled error executing task %s", name)
                result = {
                    "name": name,
                    "status": "error",
                    "message": f"Internal error: {exc}",
                }
            results.append(result)
        else:
            results.append({
                "name": task.get("name", ""),
                "status": "skipped",
                "message": f"Unknown action: {action}",
            })

    return results
